### PR TITLE
[Diagnostics] Add diagnostic when using nonexistent '**' operator

### PR DESCRIPTION
**Pull request by mininny**
This PR adds a special error explaining that `**` operator does not exist in Swift and `Foundation.pow(_:_:)` should be used instead. 

Addresses [SR-14095](https://bugs.swift.org/browse/SR-14095)
Resolved [SR-14196](https://bugs.swift.org/browse/SR-14196)

### CODE REVIEW

**Comment by varungandhi-apple** [comment]
I think there may have been some misunderstanding; the discussion in SR-14095 was about adding the operator as `@unavailable` in the standard library, instead of changing the compiler. The precedence problem that Steve pointed out would need a test (in case we don't have one already) that precedence is shadowed across modules.

So you'd have something like:

```
stdlib : defines ** with some precedence
mylibrary : defines ** with different precedence
mytest : imports mylibrary and (implicitly) stdlib, uses precedence of ** from mylibrary
```

Alternately, if we know that precedence is not shadowed, then marking the operator as unavailable is not an option (because it might break code that works today), in which case we can add a special-case diagnostic to the compiler, like you've done here.

**Comment by xwu** [comment]
I'm wary of adding a warning that something doesn't exist to the compiler or standard library when it is possible that we will want it to exist in the future.

Particularly in the case of the standard library, it buys us the disadvantages of ABI stability locking in a design that we are explicitly not yet implementing.

In the case of `**` specifically, the existing error is plenty clear: there is no such operator. I really don't see evidence that users are confused about this unambiguous message.

**Comment by varungandhi-apple** [comment]
> I'm wary of adding a warning that a specific something doesn't exist to the compiler or standard library when it is possible that we will want it to exist in the future.

We can always remove a warning/remove an `@unavailable` declaration.

> Particularly in the case of the standard library, it buys us the disadvantages of ABI stability locking in a design that we are explicitly not yet implementing.

I don't get how adding a compiler warning or adding an `@unavailable` declaration affects ABI? Could you expand on what you mean here?

> In the case of ** specifically, the existing error is plenty clear: there is no such operator. I really don't see evidence that users are confused about this unambiguous message.

It's not about confusion; it's about helpful-ness. The diagnostic is clear, but it's not immediately helpful. It's pretty likely that what the user meant was exponentiation (given the usage of `**` in other languages like Javascript/Typescript).

> The long-term solution is unambiguous, to incorporate the work in swift-numerics and actually to add an exponentiation operator.

I think that having an unambiguous long-term solution should not stop us from making incremental improvements, especially when the incremental step is not too difficult to implement (and can be easily removed at a later stage).

**Comment by xwu** [comment]
> I don't get how adding a compiler warning or adding an @unavailable declaration affects ABI? Could you expand on what you mean here?

Sorry, I am mistaken on this; the precedence group of an operator is entirely a compile time construct. However, it remains deeply problematic:

The addition of an unavailable operator here would require the addition of an exponentiation precedence group, and not breaking existing end user code inevitably relies on the [undocumented ability of end users to change the precedence of standard library operators](https://forums.swift.org/t/lets-deprecate-user-changes-to-standard-library-operator-precedence/44053). I do not think improving the diagnostic specifically for `**` is worth the tradeoff of foreclosing any possibility of warning end users about silently unintended behavior that arises when users accidentally trip on this mis-feature, which can really _only_ serve a role in the narrow circumstance where the standard library _deliberately vends a non-working operator_.

> I think that having an unambiguous long-term solution should not stop us from making incremental improvements, especially when the incremental step is not too difficult to implement (and can be easily removed at a later stage).

It's not about the difficulty of implementation; it's about the knock-on effects here which result in _promoting_ reliance on a shadowing feature that is actively harmful (see link in Swift Forum post for an actual real-world example of unintended use leading to harm).

I do agree that a compiler diagnostic is an incremental improvement in this circumstance, but I think the increment is relatively minor in that it's a nice-to-have, because in the general case users don't expect the compiler to divine what they mean when they use an undefined symbol. I think it would be reasonable to add only if we actively test that it will _not_ erroneously diagnose correct usage when either end users or a future version of Swift add their own working operator `**`.

**Comment by varungandhi-apple** [comment]
> The addition of an unavailable operator here would require the addition of an exponentiation precedence group (requires Swift Evolution or else the use of an underscored throwaway precedence)

A underscored precedence is fine IMO since we are not adding a public API to the stdlib.

> not breaking existing end user code inevitably relies on the undocumented ability of end users to change the precedence of standard library operators

No, it does not. It is relying on the ability to shadow precedence, not change that for the stdlib operator. There's a distinction between the two because `**` is not defined today by the stdlib (whereas in your linked post, the stdlib does define `-` today). Specifically, existing code that compiles today falls into one of these three buckets:

- Possibly: It defines a precedence for `**` and a `func **`. This code is unaffected because you are defining the precedence for your own operator, which will be preferred over any `@unavailable` operator introduced in the stdlib. <- [I agree that we should add a test for this](https://github.com/apple/swift/pull/35933#issuecomment-778194859). 
- Very unlikely: It defines a precedence for `**` but no `func **`. This code is unaffected because `**` was not usable before (it is not defined), and it not usable after this change (`**` is `@unavailable`).
- Not using `**` at all.

> because in the general case users don't expect the compiler to divine what they mean when they use an undefined symbol

I don't think discussing the general case is very useful here. The whole point of diagnostics is special casing, based on what problems people hit in practice.

>  I think it would be reasonable to add only if we actively test that it will not erroneously diagnose correct usage when either end users or a future version of Swift add their own working operator `**`.

Sure, I'm not opposed to have tests be added. Having solid tests is important, and I agree. That's precisely what I was getting at earlier with "would need a test (in case we don't have one already) that precedence is shadowed across modules". I think that should cover this situation.

**Comment by xwu** [comment]
> No, it does not. It is relying on the ability to shadow precedence, not change that for the stdlib operator. There's a distinction between the two because `**` is not defined today by the stdlib (whereas in your linked post, the stdlib does define `-` today). Specifically, existing code that compiles today falls into one of these three buckets:
>
> >Possibly: It defines a precedence for `**` and a func `**`. This code is unaffected because you are defining the precedence for your own operator, which will be preferred over any `@unavailable` operator introduced in the stdlib. <- I agree that we should add a test for this.

I'm sorry, I don't understand the distinction you are making. If we define a precedence group, an operator `**` assigned to that precedence group, and an unavailable `func **` in the standard library (recalling that the operator declaration and the function declaration are separate), then the "ability to shadow precedence" which you agree is required, and which exists in Swift today, _changes_ precedence for the standard library operator.

Unless--are you suggesting that as part of this diagnostic improvement we need first to implement support for operator redeclarations shadowing the standard library with different precedence while _not changing_ the precedence of the standard library operator? For a full implementation, an end user importing a third party exponentiation library would then need to be able to _choose_ which of the incompatible operator declarations is imported, for which we currently do not have a disambiguation syntax. That is a massive undertaking.

**Comment by varungandhi-apple** [comment]
(I'm writing this assuming the precedence shadowing works as described before.)

> If we define a precedence group, an operator ** assigned to that precedence group, and an unavailable func ** in the standard library (recalling that the operator declaration and the function declaration are separate), then the "ability to shadow precedence" which you agree is required, and which exists in Swift today, changes precedence for the standard library operator.

What I'm saying is, while you are technically correct, this doesn't matter, because the `**` operator isn't present in the stdlib today, and hence no one is relying on it (because it doesn't exist). Moreover, if we add it but mark it as `@unavailable`, then nobody can rely it either. So it cannot break existing code.

In contrast, with the example involving `-`, the problem arises because `-` is present in the stdlib already, which means that people are relying on it. Specifically, the failure mode you are pointing out in the forum post involves the following steps:

1. The stdlib defines an operator X with an available declaration and a particular precedence.
2. Some code starts using the operator X in their code with the stdlib's precedence.
3. Either that code, or one of its dependencies introduces a new precedence for operator X, hence potentially changing the meaning of the code written earlier.

However 1. is not applicable to the `**` case, so it will not break existing code, hence 2. doesn't follow.

From what I can tell, the worst thing that can happen is:

1. We add `**` with an underscored precedence and mark it as `@unavailable`.
2. Someone who is using the new stdlib writes a new `func **` without defining a precedence level.
3. This new `func **` is used somewhere; since the code didn't define a precedence level, the compiler picks up the underscored precedence in the stdlib.
4. If at some point, we decide to add `func **` to the stdlib with a non-underscored precedence which is different from the previous underscored precedence, that breaks code that was silently relying on the underscored precedence breaks.

Is this the failure mode you're pointing out? Thinking more about it, it does seem like a bit of a headache.

I think that makes a stronger case for preferring changing the compiler diagnostic (as @mininny has already implemented) vs changing the stdlib, which can create this hazard.

> Unless--are you suggesting that as part of this diagnostic improvement we need first to implement support for operator redeclarations shadowing the standard library with different precedence while not changing the precedence for invocations of the standard library operator? 

That's not at all what I'm suggesting.

(Sorry @mininny for flip-flapping, I need to think about this some more, but given the discussion here, I'm starting to think that your current solution might be the way to go, in which case you don't need to make more changes...)

**Comment by varungandhi-apple** [comment]
Couple of points from other folks:

> Joe: maybe we could hack it so you don't need a precedence declaration if the operator func definition is unavailable
>
> Varun: Yeah... That seemed like one option... I figured the diagnostic option is cleaner (and less work), given that this is a one-off thing... 
>
> It doesn’t have to be one-off; there might be other situations where we want to catch typos in operators. Like an =~ telling you to use ~= instead.

I wonder how hard Joe's suggestion would be to implement...

**Comment by xwu** [comment]
> I wonder how hard Joe's suggestion would be to implement...

First, a custom infix operator without a stated precedence gets the default precedence. 

Second, if you don't know the precedence of an operator, then you can't determine the type of the operands unless there are no other operators and precisely one non-literal value on the left-hand side and one of the right-hand side. (And by the same token, if you use the default precedence instead of the intended precedence, then you may infer the _wrong_ type for your operands.) And if you don't know or have incorrect information on the type of the operands, you cannot determine if you're trying to call the function that's marked `@unavailable` or some other overload (be it existent or nonexistent).

> From what I can tell, the worst thing that can happen is:
> 
> We add ** with an underscored precedence and mark it as `@unavailable`.
> Someone who is using the new stdlib writes a new func ** without defining a precedence level.
> This new func ** is used somewhere; since the code didn't define a precedence level, the compiler picks up the > underscored precedence in the stdlib.
> If at some point, we decide to add func ** to the stdlib with a non-underscored precedence which is different from the previous underscored precedence, that breaks code that was silently relying on the underscored precedence breaks.

No, that is not the worst thing that can happen. The worst thing that can happen is:

1. An end user (Alice) writes code without ever using `**`, but Alice imports an SPM package written by another person (Bob) who (for Bob's own purposes) defines `**` with non-exponentiation semantics and precedence (totally unrelated to math, let's say).  Alice doesn't know that Bob's SPM package vends an operator declaration `**` and doesn't have to, but it is used internally in Bob's library.
2. We add `**` with an underscored precedence and mark it as `@unavailable`. Everything continues to work for Alice because the behavior of Bob's code is unchanged and his library continues to compile just fine on Alice's machine, since Bob is allowed to shadow the precedence of `**`.
3. We decide to add an actual implementation of `**` with a non-underscored precedence which is different from Bob's library. Everything still works fine, thanks to shadowing.
4. Alice decides to use the new exponentiation operator when she modernizes her code to take advantage of the new features in Swift Next. Alice's code silently produces different results because the precedence of the standard library operator `**` relative to the other standard library arithmetic operators has been altered by Bob's library, even though Alice never uses Bob's `**` and doesn't even know it exists, and even though Bob's library has absolutely nothing to do with math.
5. We can't fix this problem for Alice by banning Bob's library from changing the precedence of operators declared in the standard library, because we've been relying on that mis-feature precisely to do what we did in steps 2 and 3 without breaking Bob's code from step 1.

This is double plus ungood.

**Comment by varungandhi-apple** [comment]
> If you don't know the precedence of an operator, [...]

I don't think this is a problem in practice because the `@unavailable` operator can be marked to have a very general type `<T, U> (T, U) -> T`, which should catch anything that falls through. Here's what I tried:

```
// added to stdlib locally
infix operator **: DefaultPrecedence
@available(*, unavailable, message: "Use Foundation.pow(_:) for exponentation")
public func **<T, U>(_ t: T, _: U) -> T { fatalError() }

// test.swift
let x1 = 4.0 + 3.0 ** 2.0
// error: adjacent operators are in unordered precedence groups 'AdditionPrecedence' and 'DefaultPrecedence'
// error: '**' is unavailable: Use Foundation.pow(_:) for exponentation

func **(_ : Int, _: Int) -> Int { 0 }
let y = 3 ** 2 // OK
```

It's a bit annoying to have the 'unordered precedence groups' error... Maybe we should treat `@available(*, unavailable)` operators as implicitly having the highest precedence so that particular error is not emitted...

> No, that is not the worst thing that can happen. The worst thing that can happen is: [...]

Couple of points:

1. If we special-case unavailable operators to have a precedence that does not participate in shadowing, this is a non-issue.
2. The failure mode you've pointed out here is something that is not specific to operators, it applies generally to anything we add to the stdlib, including the case where we add a new operator as properly available. Due to the presence of pervasive shadowing facilities in the language, any person writing code needs some awareness of what declarations are being called in a particular context. Are you implying that we should not add any operators to the stdlib until we add a warning for situations where precedence is getting shadowed?

---

Based on the discussion so far, my current thinking is that:

1. Unavailable operators should get an artificially high precedence, to avoid superfluous "unordered precedence groups" errors.
2. This artificial high precedence cannot be shadowed (it's not really a declaration, it's an implementation detail of the compiler), so any downstream modules still need their own precedence declaration.

**Comment by xwu** [comment]
> It's a bit annoying to have the 'unordered precedence groups' error... Maybe we should treat `@available(*, unavailable)` operators as implicitly having the highest precedence so that particular error is not emitted...

(1) Two operators can be unordered relative to each other regardless of their relative precedence, because we overhauled the system of precedence precisely to allow this; you'd have to stipulate that this "implicit highest precedence group" is also ordered relative to every other precedence group wherever defined.

(2) The point here is that you'd have to parse the operator with _some_ precedence, and if it's not the right one, you've changed the type of the resulting expression. It does not suffice to have a generic `func **<T, U>(_ t: T, _: U) -> T`, because ultimately that means that you're going to have a type `T` as the argument for the next operator (in your example, as the right-hand side of `+`), and if `T` is incorrect because you got the precedence incorrect, then your typechecking for `+` is going to be incorrect, and that will have knock-on effects everywhere.

> If we special-case unavailable operators to have a precedence that does not participate in shadowing, this is a non-issue.

I don't understand: Are you suggesting that we create a special precedence group that also specifically forbids shadowing of operators that have that precedence by compiler magic?

> The failure mode you've pointed out here is something that is not specific to operators, it applies generally to anything we add to the stdlib, including the case where we add a new operator as properly available.

The failure mode doesn't apply in the same way because:
(1) Operators cannot be disambiguated by prefixing a module name but are in scope whenever a module is imported, causing the shadowing you describe above that affects parsing of functions not defined in the same module; no other feature in the language exhibits this behavior, as far as I'm aware.
(2) It is arguably justifiable to have a compiler warning that says: "Hey, the standard library now has `**` as the exponentiation operator, so you should be aware that you're shadowing it with a different precedence, because your code could behave strangely." This is precisely what I want to add. It's _not_ justifiable to have such a warning if we use this (mis-)feature for the diagnostics proposed here, because where we do this, any such warning instead means, "Hey, the standard library isn't introducing `**`, but we're going to warn you away from filling that gap yourself as well; it's for your own good, so that you know that we're haven't introduced `**`."

I really, really urge this warning--if it's added at all--to be a compiler diagnostic.

**Comment by xwu** [comment]
What's to stop us, incidentally, from actually just adding `ExponentiationPrecedence` and `**` into the standard library _right now_ (after going through the right process in Evolution)? Foundation could then choose to add `@_alwaysEmitIntoClient func **`, with an eye to sinking the function into the standard library once swift-numerics features are incorporated into the standard library. This would require no new features and no diagnostics, because things would _just work_.

**Comment by varungandhi-apple** [comment]
> "implicit highest precedence group" is also ordered relative to every other precedence group wherever defined, which can only happen by compiler magic.

Yes, that's what I was thinking. It would an implementation detail inside the compiler (simply as a mechanism to suppress the unhelpful diagnostic).

> as if it's not the right one, you've changed the type of the resulting expression [..]

If the expression is ill-formed (due to using a non-existing operator being used), I don't think it matters what type it gets, so long as we (a) provide a helpful diagnostic and (b) minimize non-helpful diagnostics.

> I really, really urge this note about how to exponentiate--if it's added at all--to be a compiler diagnostic.

Could you clarify if you've changed your perspective from before (I thought you were opposed to both the diagnostic and the stdlib approaches)?

> It's not justifiable to have such a warning **if we use this (mis-)feature** (emphasis added) for the diagnostics proposed here, because where we do this any such warning instead means, "Hey, the standard library isn't introducing **, but we're going to warn you away from filling that gap yourself as well; it's for your own good, so that you know that we're haven't introduced **."

That's why I mentioned "This artificial high precedence cannot be shadowed (it's not really a declaration, it's an implementation detail of the compiler), so any downstream modules still need their own precedence declaration.", so we are very much NOT using this "(mis-)feature".

Maybe I'm misunderstanding your point, but I don't really see how the two are in opposition. It would be different if the code behaved as-if the operator was declared with the very high precedence in the source (which would entail the dependence on shadowing behavior that you pointed out), but that's not what I'm proposing.

> What's to stop us, incidentally, from actually just adding [...]

I'm not sure if this a rhetorical question or not... Going through evolution is a whole process with writing a proposal etc. Changing a compiler diagnostic is much simpler and addresses the immediate problem...

4ykvcs


---

Perhaps it is worth re-iterating / clarifying what I'm proposing:

1. Make some changes to the compiler so that unavailable operators get special treatment. This involves:
   a. A compiler-internal precedence higher than all other precedences. (This avoids the unhelpful diagnostic)
   b. The restriction to assign a precedence group in the source is lifted.
   c. This precedence cannot be shadowed (unlike a hand-written precedence in the source code as I did earlier). This means that if any library defines that operator, that library is required to assign a precedence group for that operator.
2. Add unavailable operator `**` with a message

In the future, we can potentially:
- Add the warning that you suggested, if someone can come up with an implementation and get some consensus that it is an okay warning to add. (Personally, I don't have strong opinions on this.)
- Add `=~` as unavailable to the stdlib, as suggested by Becca.
- Add `**` to the stdlib (after going through evolution).

**Comment by xwu** [comment]
> If the expression is ill-formed (due to using a non-existing operator being used), I don't think it matters what type it gets, so long as we (a) provide a helpful diagnostic and (b) minimize non-helpful diagnostics.

It would cause potentially incorrect diagnostics about the use of `+`, etc., because the operands do not have the types intended--i.e., non-helpful diagnostics that could be actively misleading.

> I'm not sure if this a rhetorical question or not... Going through evolution is a whole process with writing a proposal etc. Changing a compiler diagnostic is much simpler and addresses the immediate problem...

One does not simply...add magical behavior just to avoid Swift Evolution. In that this would have knock-on user-visible effects other than diagnostics, it would arguably require Swift Evolution as well. An Evolution proposal is not so onerous as to be impossible; I am happy to draft it _tonight_.

> Make some changes to the compiler so that unavailable operators get special treatment. This involves:
> a. A compiler-internal "peak precedence" higher than all other precedences. (This avoids the unhelpful diagnostic)
> b. The restriction to assign a precedence group in the source is lifted (only for unavailable operators); if it not defined, then the peak precedence is used.
> c. This precedence cannot be shadowed (unlike a hand-written precedence in the source code as I did earlier). This means that if any library defines that operator, that library is required to assign a precedence group for that operator.

Please, do not do this; I have tried very hard to explain why it would have very bad effects. I'm not sure what more I can say on that. Everything else is...fine, just not this.

**Comment by varungandhi-apple** [comment]
Okay, stepping back a bit, say we discarded the proposal about the high precedence etc. Let's put it away.

Are you okay with a special-case compiler diagnostic being added, as the current state of the PR stands (we can tweak the exact diagnostic wording)? Or are you against that (and suggesting that we should instead make a Swift evolution proposal)?

**Comment by xwu** [comment]
> Are you okay with a special-case compiler diagnostic being added, as the current state of the PR stands (we can tweak the exact diagnostic wording)? Or are you against that (and suggesting that we should instead make a Swift evolution proposal)?

I'd be OK, but having an unavailable operator allows us to emit even more targeted diagnostics (when the type is appropriate, etc.) and I think we can do that _very_ soon even if it requires Evolution :)

**Comment by varungandhi-apple** [comment]
OK, given that, I think let's tweak the wording/PR a little bit and land this change, since it is mostly ready.

When the evolution proposal is approved, we can potentially go ahead and remove/revert this.

File: lib/Sema/PreCheckExpr.cpp
Comment (varungandhi-apple) [inline lib/Sema/PreCheckExpr.cpp:250]: Couple of minor notes for improvement here: (based on combined feedback from Joe and Becca)

`pow` is actually defined in the C library, and Foundation re-exports it, so maybe suggesting Foundation is not the right thing to do. Instead, we can do an unqualified lookup using `getModuleScopeContext()` (to get the scope corresponding to the current file) and the `TypeChecker::lookupUnqualified` method, to check if there is a `pow` in scope. If `pow` is in scope, we can issue an error like `no operator '**' is defined; did you mean 'pow(_:_:)'?`.

Otherwise, we don't issue the special error.

For test the diagnostic, I think you should be able to test this using either (1) the fake SDK that we have `%clang-importer-sdk` and have that provide a `pow` function OR (2) creating a small test case where one module exports `pow` and another module that imports the first one tries to use `**` without defining it.

**Comment by xwu** [comment]
> OK, given that, I think let's tweak the wording/PR a little bit and land this change, since it is mostly ready.

Sure. The Evolution proposal draft is done:
https://gist.github.com/xwu/93cd21d878889fdb0b728386b9e2dc5a

**Comment by varungandhi-apple** [comment]
Reading your proposal, I realize my mistake:

>  The addition of actual exponentiation facilities is covered by SE-246, and is deliberately left out of the scope of this proposal.

When I originally read your comment, I misinterpreted the scope of what you were suggesting. I thought it would require an implementation, potential discussion on adding a power operation on integers, and float with integer exponents etc... 

That's kinda' why I responded the way I did. I also didn't know about SE-246. So your proposal is much smaller because of that. Sorry about misreading your comment and the resulting back-and-forth.

**Comment by xwu** [comment]
No worries, thanks for reading; yeah, the changes are teeny-tiny.

File: lib/Sema/PreCheckExpr.cpp
Comment (mininny) [reply lib/Sema/PreCheckExpr.cpp]: Thanks! I'm not sure if this can be merged since @xwu already created a proposal, but I changed it to lookup for `pow` as you suggested :)

File: test/decl/operator/power_operator_imported.swift
Comment (varungandhi-apple) [inline test/decl/operator/power_operator_imported.swift:12]: Can you add one more test case where there is no `pow` in scope, and in that situation check that we are not emitting the specialized diagnostic? I'm not sure if `-verify` has a way to do it, but you can pipe the output to FileCheck to do so.

Maybe it can be part of this same file but with 2 RUN lines `-DWITH_POW`, `-DWITHOUT_POW` or something.

File: lib/Sema/PreCheckExpr.cpp
Comment (varungandhi-apple) [reply lib/Sema/PreCheckExpr.cpp]: > I'm not sure if this can be merged [..]

I think your PR is already in good shape and provides a nice improvement, so we should merge it. From a cursory reading, the discussion on the forums seems to be a bit animated (different people are pointing out that different behavior is desirable), and then it will also need to be scheduled for review + core team approval. I'm guessing the proposal should be accepted, but I think we should merge this in case that doesn't pan out.

Comment (xwu) [reply lib/Sema/PreCheckExpr.cpp]: I agree; I certainly have no objections to trying to improve the diagnostics here. I like @varungandhi-apple's suggestions above.

File: include/swift/AST/DiagnosticsSema.def
Comment (xwu) [inline include/swift/AST/DiagnosticsSema.def:847]: "Unexisting" is not an English word. Do you mean "nonexistent"?

File: lib/Sema/PreCheckExpr.cpp
Comment (xwu) [inline lib/Sema/PreCheckExpr.cpp:241]:
```suggestion
static bool diagnoseUnexistingOperators(DiagnosticEngine &Diags,
```

Same comment here. Also, since this method only diagnoses "**", perhaps that could be reflected in the name; it diagnoses no other operators.

File: lib/Sema/PreCheckExpr.cpp
Comment (xwu) [inline lib/Sema/PreCheckExpr.cpp:455]: See this line for an example of the name of a function that diagnoses a specific operator.

File: test/decl/operator/power_operator_imported.swift
Comment (mininny) [reply test/decl/operator/power_operator_imported.swift:12]: New test in `operator.swift` test does that where no `pow` is in scope and only generic `cannot find operator '**' in scope` error is emitted.

File: include/swift/AST/DiagnosticsSema.def
Comment (mininny) [reply include/swift/AST/DiagnosticsSema.def:847]: Thanks! I changed it to `nonexistent_power_operator`

File: lib/Sema/PreCheckExpr.cpp
Comment (mininny) [reply lib/Sema/PreCheckExpr.cpp:241]: Thanks :) I modified it to `diagnoseNonexistentPowerOperator`

**Review by varungandhi-apple — APPROVED** [review]
Thank you! LGTM.

**Comment by varungandhi-apple** [comment]
@swift-ci smoke test

File: test/decl/func/operator.swift
Comment (xwu) [inline test/decl/func/operator.swift:426]:
```suggestion
func testNonexistentPowerOperatorWithPowFunctionOutOfScope() {
```

Something like that, maybe? The name doesn't make sense to me otherwise.

File: test/decl/operator/Inputs/power_operator_declaration/PowerOperatorDeclaration.h
Comment (xwu) [inline test/decl/operator/Inputs/power_operator_declaration/PowerOperatorDeclaration.h:1]: I'm not sure I understand: Why is this Obj-C module called a "power operator declaration"? It doesn't seem to declare a power operator anywhere. Do you mean simply that it declares `pow` (not the operator)? Can we just test by importing the C standard library directly instead?

File: test/decl/operator/declared_power_operator.swift
Comment (xwu) [inline test/decl/operator/declared_power_operator.swift:7]:
```suggestion
func testPowerOperator() { 
```

I would not go so far as to call it a _valid power operation_: it does no such thing :)

File: test/decl/operator/power_operator_imported.swift
Comment (xwu) [inline test/decl/operator/power_operator_imported.swift:7]:
```suggestion
func testNonexistentPowerOperatorWithPowFunctionInScope() {
```

How about something like that?

**Review by xwu — COMMENTED** [review]
Some naming suggestions and formatting nits.

I'd be careful about naming things that define an _operation_ from those that define an _operator_.

The _operator_ is `**`. It can _exist_ or _not exist_. I don't know how to parse it when you call an operator "valid" or "invalid," as there are no tests here about invalid declarations.

The _operation_ is exponentiation. A _valid_ operation is one that works (i.e., exponentiates), and an _invalid_ one that doesn't; you're not testing any of that. You're running tests about diagnostics behavior based on whether the operation is _in scope_ or _out of scope_.

File: test/decl/operator/Inputs/power_operator_declaration/PowerOperatorDeclaration.h
Comment (mininny) [reply test/decl/operator/Inputs/power_operator_declaration/PowerOperatorDeclaration.h]: Thanks I removed it and fixed it to test by importing `cfuncs` instead :)

File: test/decl/operator/power_operator_imported.swift
Comment (mininny) [reply test/decl/operator/power_operator_imported.swift]: Thanks! I modified all function names as you suggested! 👍

File: test/decl/operator/declared_power_operator.swift
Comment (xwu) [inline test/decl/operator/declared_power_operator.swift:7]: Thanks--can we simplify this name? It seems redundant to me. `testPowerOperator` or `testDeclaredPowerOperator` seems fine.

Comment (mininny) [reply test/decl/operator/declared_power_operator.swift:7]: Fixed! :)

**Comment by xwu** [comment]
@swift-ci smoke test

**Comment by mininny** [comment]
Can this be merged? :)

**Comment by xwu** [comment]
Looks good. Can we add one more test? When `**` is declared and `func **` for floating point, let’s test what happens when users try to use the operator on two `Int8` values.

**Comment by mininny** [comment]
> Looks good. Can we add one more test? When `**` is declared and `func **` for floating point, let’s test what happens when users try to use the operator on two `Int8` values (without an implementing function—basically, testing when there’s an operator declaration but no function).

It emits `cannot convert value of type 'Int8' to expected argument type 'Double'` which seems correct. Right?

**Comment by xwu** [comment]
Great, that works! Let’s just make sure there’s a test for it.

**Comment by mininny** [comment]
> Great, that works! Let’s just make sure there’s a test for it.

Done! :)

**Comment by xwu** [comment]
@swift-ci please smoke test

File: lib/Sema/PreCheckExpr.cpp
Comment (xedin) [inline lib/Sema/PreCheckExpr.cpp:242]: Nit: Formatting here needs to be adjusted, please use clang-format or git-clang-format to fix it.

File: lib/Sema/PreCheckExpr.cpp
Comment (xedin) [inline lib/Sema/PreCheckExpr.cpp:245]: You can use `if (!(name.isOperator() && name.is("**"))` condition to avoid nesting.

**Review by xedin — APPROVED** [review]
Looks good to me, I have left a couple of inline comments and code and formatting.

**Comment by mininny** [comment]
> Looks good to me, I have left a couple of inline comments and code and formatting.

Thanks! I fixed the formatting and removed nesting :)

**Comment by xwu** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

**Pull request merged**
